crates/pathfinder/src/state/sync/pending.rs

```diff
@@ -28,60 +28,83 @@ pub async fn poll_pending(
     use pathfinder_common::BlockId;
     use std::sync::Arc;
 
+    let mut prev_block = Arc::new(PendingBlock {
+        gas_price: Default::default(),
+        parent_hash: Default::default(),
+        sequencer_address: Default::default(),
+        status: starknet_gateway_types::reply::Status::Pending,
+        timestamp: Default::default(),
+        transaction_receipts: Default::default(),
+        transactions: Default::default(),
+        starknet_version: Default::default(),
+    });
+
+    let mut prev_state_update = Arc::new(StateUpdate::default());
+
     loop {
         use starknet_gateway_types::reply::MaybePendingBlock;
 
-        let block = match sequencer
-            .block(BlockId::Pending)
-            .await
-            .context("Download pending block")?
-        {
-            MaybePendingBlock::Block(block) if block.block_hash == head.0 => {
-                // Sequencer `pending` may return the latest full block for quite some time, so ignore it.
-                tracing::trace!(hash=%block.block_hash, "Found current head from pending mode");
-                tokio::time::sleep(poll_interval).await;
-                continue;
-            }
-            MaybePendingBlock::Block(block) => {
-                tracing::trace!(hash=%block.block_hash, "Found full block, exiting pending mode.");
-                return Ok((Some(block), None));
+        let block = loop {
+            let block = match sequencer
```

Comment:
The only thing I don't like about this change is that here we fall into a retry loop without any kind of delay. Are we sure that's not a problem?

It still looks strange that in the outer loop we have the sleep with poll_interval but here we're pounding the gateway as hard as we can.

Reply:
I think you're right, I was trying to avoid rewriting this too much. But I think its probably better to restructure this.

Will change to draft and work a bit more.

Reply:
I've refactored it entirely. See if its better now..